ddtrace/appsec/_processor.py

```diff
@@ -101,16 +101,22 @@ def get_appsec_obfuscation_parameter_value_regexp() -> bytes:
     "x-real-ip",
 }
 
+_COLLECTED_REQUEST_HEADERS_ASM_ENABLED = {
+    "accept",
+    "content-type",
+    "user-agent",
+}
+
 
-def _set_headers(span: Span, headers: Any, kind: str) -> None:
+def _set_headers(span: Span, headers: Any, kind: str, only_asm_enabled: bool = False) -> None:
     from ddtrace.contrib.trace_utils import _normalize_tag_name
 
     for k in headers:
         if isinstance(k, tuple):
             key, value = k
         else:
             key, value = k, headers[k]
-        if key.lower() in _COLLECTED_REQUEST_HEADERS:
+        if key.lower() in (_COLLECTED_REQUEST_HEADERS_ASM_ENABLED if only_asm_enabled else _COLLECTED_REQUEST_HEADERS):
```

Comment:
I'm not sure I get it, isn't the first a subgroup of the second? So if ASM is enabled we're sending less headers?

Reply:
Nothing is done if asm disabled in that file. But previously we called set_headers only if a waf event was detected. Now we call set_headers on request at the end of the span with "only_asm_enabled" to collect those headers that must be always collected if asm is enabled. But we must not collect all the headers at that point.